### ISSUE TITLE
Expert Syncenet  training?

### ISSUE THREAD

**Issue by rizwanishaq** [Rudrabha/Wav2Lip#195]
I am trying to train the expert syncnet on LRS3 Data set but loss is stuck around 0.6 from the start and not changing much? any idea?

**Comment by prajwalkr** [comment]
The syncnet only starts converging below 0.69 after about 150K iterations on LRS2.

**Comment by rizwanishaq** [comment]
Ok. I am doing everything with tensorflow, and developed the model accordingly. If you want we can start new tensorflow branch.

**Comment by wangzheliu** [comment]
> 
> 
> I am trying to train the expert syncnet on LRS3 Data set but loss is stuck around 0.6 from the start and not changing much? any idea?

How long  time does one epoch spend？Thank you

**Comment by rizwanishaq** [comment]
As I am training on two GPU (with 128 Batch size) and training it on trainval (LRS3), so each epoch from me it is taking around 29 seconds.

**Comment by rizwanishaq** [comment]
Beside that, I also changed the data pipeline little bit, such as,

1. Using ([RetinaNetMobileNetV1](https://pypi.org/project/face-detection/)) for face detection (270ms (3.7 FPS) on cpu)
2. Pure[ Tensorflow implementation](https://www.tensorflow.org/io/tutorials/audio) of mel-spectrum instead of librosa 
3. From every video I am take both positive and negative pair (only with audio)

**Comment by rizwanishaq** [comment]
It is so strange, that even after two days of training, syncnet is still stuck at 0.69?

**Comment by prajwalkr** [comment]
> It is so strange, that even after two days of training, syncnet is still stuck at 0.69?

Convergence happens only after 200K iterations or so

**Comment by rizwanishaq** [comment]
@prajwalkr I changed the learning rate to 5e-5 and beta_1=0.5 in adam now.. the sync model reached to 0.2 so fast, i.e. with 6 GPU and 128 batch size, it reached to 0.2 within 20 epochs(i.e 1 hrs) , no idea is it normal behaviour???

**Comment by prajwalkr** [comment]
That's great! Is it on LRS2?

**Comment by rizwanishaq** [comment]
It is on LRS2 dataset, I am checking now with LRS3 dataset as well. Let you know soon.

**Comment by wangzheliu** [comment]
> 
> 
> That's great! Is it on LRS2?
I train  the expert syncnet ，python color_syncnet_train.py,but it shows 
/anaconda3/envs/wav2lip/lib/python3.6/site-packages/librosa/core/spectrum.py:22
4: UserWarning: n_fft=800 is too small for input signal of length=683
Also although it is a userwarning.I used dataest are 1~3 Seconds videos like LRS2

**Comment by rizwanishaq** [comment]
@wangzheliu warning make sense, as the signal length is less than window size which is equal to n_fft=800?

**Comment by wangzheliu** [comment]
> 
> 
> @wangzheliu warning make sense, as the signal length is less than window size which is equal to n_fft=800

Thank you very much,its means some videos are too short?

**Comment by prajwalkr** [comment]
> its means some videos are too short?

I don't think videos in LRS2 are that short. Most likely your downloaded/extracted videos are corrupted.

**Comment by wangzheliu** [comment]
> 
> 
> > its means some videos are too short?
> 
> I don't think videos in LRS2 are that short. Most likely your downloaded/extracted videos are corrupted.

Thank you for you reply,I delete 1~2 videos, warning has disappear.

**Comment by ascust** [comment]
@rizwanishaq  Hey man, any updates on LRS3? I met the same issue as yours. stuck at 0.69.

**Comment by rizwanishaq** [comment]
@ascust @prajwalkr Yes, I reached around 0.4 until now .... but, I am training in tensorflow, not with pytorch... I hope @prajwalkr we can start tensorflow branch soon....

**Comment by ascust** [comment]
@rizwanishaq how many iters did you spend to get 0.4?

**Comment by rizwanishaq** [comment]
@ascust I reached with 30 epochs with (128 Batch size, and 2 GPU), I keep training more, so lets see where it reached... I am using pretrain (dataset for training) and test for validation. I don't know, how good this test dataset is.

**Comment by rizwanishaq** [comment]
@ascust @prajwalkr On LRS2 dataset (train/val for training, test for validation) I achieved validation loss of 0.14.

**Comment by Mayur28** [comment]
@prajwalkr @Rudrabha I'm not sure why, but I have the opposite problem where using the code from the repo as is, the Loss is going below 0.2 in less than 100k steps. The evaluation loss is also continuously increasing. Has anyone experienced this before? I've attached a plot of the loss vs steps.
![image](https://user-images.githubusercontent.com/26574827/124107026-1a233880-da65-11eb-8385-762a65d1fcc7.png)

Thanks!

**Comment by i-amgeek** [comment]
> @rizwanishaq Hey man, any updates on LRS3? I met the same issue as yours. stuck at 0.69.

@ascust I am also stuck at 0.69 on LRS3. Have you been able to solve your problem? If yes, can you please tell me what were the changes you had to make?

**Comment by ascust** [comment]
@i-amgeek Sorry man, I did not really solve the problem. I tried to increase the batch size and changed the sampling method (balance pos and neg samples in each batch) and the loss got to decrease to a lower level (around 0.3). I gave up after that.

**Comment by nhutsam261** [comment]
> .

I did the same but the result is not converged anymore, just stays at 0.69 ~ 0.7

**Comment by DDDlk** [comment]
> @prajwalkr @Rudrabha I'm not sure why, but I have the opposite problem where using the code from the repo as is, the Loss is going below 0.2 in less than 100k steps using LRS2. The evaluation loss is also continuously increasing. Has anyone experienced this before? I've attached a plot of the loss vs steps. ![image](https://user-images.githubusercontent.com/26574827/124107026-1a233880-da65-11eb-8385-762a65d1fcc7.png)
> 
> Thanks!

hello, which tools u used for visualization?

**Comment by Mayur28** [comment]
Hi @DDDlk ,

I have been using _Weights & Biases_.

**Comment by DDDlk** [comment]
> Hi @DDDlk ,
> 
> I have been using _Weights & Biases_.
Thank u very much~

**Comment by DDDlk** [comment]
> Hi @DDDlk ,
> 
> I have been using _Weights & Biases_.

Thank u very much~

**Comment by xiao-keeplearning** [comment]
> @prajwalkr I changed the learning rate to 5e-5 and beta_1=0.5 in adam now.. the sync model reached to 0.2 so fast, i.e. with 6 GPU and 128 batch size, it reached to 0.2 within 20 epochs(i.e 1 hrs) , no idea is it normal behaviour???

Set **beta_1=0.5** !!!  As far as I know, such a setting is extremely rare, can you explain the reason for this?

**Comment by weisrc** [comment]
Hi, for this model, is it a good idea to add gradient clipping? Right now I set it to 1 and the model seems to remain at a loss of 0.69 after 3k epochs with a batch size of 256 (3 iterations per epoch). I trained for 20 hours. Perhaps, is the batch size too big for this?
